File: stack/stack.rb
require "pry"

class Stack
  attr_accessor :first, :n

  def initialize
    @first = nil
    @n = 0  # number of elements in the Stack
  end

  class Node
    attr_accessor :item, :next

    def initialize
      @item = 0
      @next = nil
    end
  end

  def is_empty?
    @n == 0
  end

  def size
    @n
  end

  def push(item)
    old_first = @first
    @first = Node.new
    @first.item = item
    @first.next = old_first
    @n += 1
  end

  def pop
    # Remove first item from the list
    item = @first.item
    @first = first.next
    @n -= 1
    item
  end
end

binding.pry

puts "cookie"

File: stack/stack.js
class Stack {
  constructor() {
    this.first = null;
    this.size = 0;
  }

  isEmpty() {
    return this.first === null;
  }

  sizeOfStack() {
    return this.size;
  }

  push(item) {
    if(this.first === null) {
      this.first = item;
      this.size++;
      return
    }

    let old_first = this.first;
    let node = new Node(item);
    node.next = old_first;
    this.first = node;
    this.size++;
    return
  }

  pop() {
    if(this.first === null) { return }
    let item = this.first;
    this.first = this.first.next;
    this.size--;
    return item;
  }
}

class Node {
  constructor(value){
    this.next = null;
    this.value = value;
  }
}

let stack = new Stack();
stack.isEmpty();
stack.push('A');
stack.push('B');
stack.push('C');
stack.pop();
stack.sizeOfStack();

